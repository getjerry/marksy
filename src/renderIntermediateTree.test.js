/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/jsx-one-expression-per-line */
import React, { createElement, Component } from 'react';
import PropTypes from 'prop-types';
import Preact from 'preact';
import preactRenderToString from 'preact-render-to-string';
import { renderToString as infernoRenderToString } from 'inferno-server';
import { createElement as infernoCreateElement } from 'inferno-create-element';
import { render } from '@testing-library/react';

import hljs from 'highlight.js/lib/highlight';
import hljsJs from 'highlight.js/lib/languages/javascript';
import hljsXml from 'highlight.js/lib/languages/xml';

// eslint-disable-next-line
import marksy, { compileToIntermediateTree, renderIntermediateTree } from './';

hljs.registerLanguage('javascript', hljsJs);
hljs.registerLanguage('xml', hljsXml);

// eslint-disable-next-line
class TestComponent extends Component {
  render() {
    return <div>{this.props.children}</div>;
  }
}

TestComponent.propTypes = {
  children: PropTypes.node,
};

TestComponent.defaultProps = {
  children: null,
};

function getCompiled(content, options, context) {
  const compile = marksy(options || { createElement });
  const compiled = compile(content, {}, context);
  return compiled;
}

function getRenderedContainer(content, options, context) {
  const compiled = getCompiled(content, options, context);
  const { container } = render(<TestComponent>{compiled.tree}</TestComponent>);
  return container;
}

function getCompiledTwoStep(content, options, context) {
  const compiled = renderIntermediateTree(
    compileToIntermediateTree(content),
    options || { createElement },
    context
  );
  return compiled;
}

function getRenderedContainerWithTwoSteps(content, options, context) {
  const compiled = getCompiledTwoStep(content, options, context);
  const { container } = render(<TestComponent>{compiled.tree}</TestComponent>);
  return container;
}

function getTwoCompiledContainer(content) {
  const container = getRenderedContainer(content);
  const containerTwo = getRenderedContainerWithTwoSteps(content);
  return [container, containerTwo];
}

describe('Two step render result should be same with marksy', () => {
  it('text', () => {
    const [container, containerTwo] = getTwoCompiledContainer('hello');
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('strong text', () => {
    const [container, containerTwo] = getTwoCompiledContainer('hello **there**');
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('italic text', () => {
    const [container, containerTwo] = getTwoCompiledContainer('hello *there*');
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('links', () => {
    const [container, containerTwo] = getTwoCompiledContainer('[my link](http://example.com)');
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('headers', () => {
    const [container, containerTwo] = getTwoCompiledContainer(`
# header1
## header2
### header3
#### header4
  `);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('should handle same name nested headers', () => {
    const [container, containerTwo] = getTwoCompiledContainer(`
# header1
## header2
# header3
## header2
  `);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('ordered list', () => {
    const [container, containerTwo] = getTwoCompiledContainer(`
1. foo
2. bar
  `);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('list', () => {
    const [container, containerTwo] = getTwoCompiledContainer(`
- foo
- bar
  `);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('tables', () => {
    const [container, containerTwo] = getTwoCompiledContainer(`
| Tables        | Are           | Cool  |
| ------------- |:-------------:| -----:|
| col 3 is      | right-aligned | $1600 |
| col 2 is      | centered      |   $12 |
| zebra stripes | are neat      |    $1 |
  `);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('codespans', () => {
    const [container, containerTwo] = getTwoCompiledContainer('install with `$ npm install`');
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('image', () => {
    const [container, containerTwo] = getTwoCompiledContainer('![test](http://some.com/image.png)');
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('html', () => {
    const [container, containerTwo] = getTwoCompiledContainer('<div>hello</div>');
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('multiple html', () => {
    const [container, containerTwo] = getTwoCompiledContainer(
      '<div>hello</div>\n<strong>there</strong>\n<em>world</em>'
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('components using H and marksy language', () => {
    const options = {
      createElement,
      components: {
        Test() {
          return <div>mip</div>;
        },
      },
    };
    const content = '```marksy\nh(Test)\n```';

    const container = getRenderedContainer(content, options);
    const containerTwo = getRenderedContainerWithTwoSteps(content, options);

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('nested lists', () => {
    const [container, containerTwo] = getTwoCompiledContainer(`
- Colors
    - Red
    - Blue
- Shape
  - Triangle
  - Rectangle
  `);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('should be able to combine in compilation', () => {
    const [container, containerTwo] = getTwoCompiledContainer(`
# hey

- foo

-bar
  `);
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('should produce TOC', () => {
    const [container, containerTwo] = getTwoCompiledContainer(`
# foo

## bar

### baz
  `);

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('should produce custom tags', () => {
    const options = {
      createElement,
      elements: {
        h1(props) {
          return <div>{props.children}</div>;
        },
      },
    };
    const content = `
# foo
  `;

    const container = getRenderedContainer(content, options);
    const containerTwo = getRenderedContainerWithTwoSteps(content, options);

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('should work with Preact', () => {
    const options = {
      createElement: Preact.h,
      elements: {
        h1(props) {
          return Preact.h('div', null, props.children);
        },
      },
    };
    const content = `
# foo
  `;

    const compiled = getCompiled(content, options);
    const compiled2 = getCompiledTwoStep(content, options);

    expect(preactRenderToString(Preact.h('div', null, compiled.tree))).toMatchSnapshot();
    expect(preactRenderToString(Preact.h('div', null, compiled.tree))).toStrictEqual(
      preactRenderToString(Preact.h('div', null, compiled2.tree))
    );
  });

  it('should work with Inferno', () => {
    const options = {
      createElement: infernoCreateElement,
      elements: {
        h1(props) {
          return infernoCreateElement('div', null, props.children);
        },
      },
    };
    const content = `
# foo
  `;

    const compiled = getCompiled(content, options);
    const compiled2 = getCompiledTwoStep(content, options);

    expect(
      infernoRenderToString(infernoCreateElement('div', null, compiled.tree))
    ).toMatchSnapshot();
    expect(infernoRenderToString(infernoCreateElement('div', null, compiled.tree))).toStrictEqual(
      infernoRenderToString(infernoCreateElement('div', null, compiled2.tree))
    );
  });

  it('should allow injecting context to elements', () => {
    const options = {
      createElement,
      elements: {
        h1(props) {
          return <div>{props.context.foo}</div>;
        },
      },
    };
    const content = `
# foo
  `;

    const context = {
      foo: 'bar',
    };

    const container = getRenderedContainer(content, options, context);
    const containerTwo = getRenderedContainerWithTwoSteps(content, options, context);

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('should allow overriding inline code element', () => {
    const options = {
      createElement,
      elements: {
        codespan({ children }) {
          return <span>{children}</span>;
        },
      },
    };
    const content = 'Hello `code`';

    const container = getRenderedContainer(content, options);
    const containerTwo = getRenderedContainerWithTwoSteps(content, options);

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('should allow overriding block code element', () => {
    const options = {
      createElement,
      elements: {
        code({ language, code }) {
          return (
            <div>
              {language}:{code}
            </div>
          );
        },
      },
    };
    const content = '```js\ncode\n```';

    const container = getRenderedContainer(content, options);
    const containerTwo = getRenderedContainerWithTwoSteps(content, options);

    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('should escape code when no highlighting is supplied', () => {
    const [container, containerTwo] = getTwoCompiledContainer(
      '```js\nconst Foo = () => <div/>\n```'
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });

  it('self-closing tag', () => {
    const [container, containerTwo] = getTwoCompiledContainer(
      `
  ![test](http://some.com/image.png)
  
  <div><br /></div>
  
  <hr/>

  <input type='text' />
  `
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toStrictEqual(containerTwo.firstChild);
  });
});
