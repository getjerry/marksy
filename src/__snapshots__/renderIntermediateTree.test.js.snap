// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Two step render result should be same with marksy codespans 1`] = `
<div>
  <p>
    install with 
    <code>
      $ npm install
    </code>
  </p>
</div>
`;

exports[`Two step render result should be same with marksy components using H and marksy language 1`] = `
<div>
  <div>
    mip
  </div>
</div>
`;

exports[`Two step render result should be same with marksy headers 1`] = `
<div>
  <h1
    id="header1"
  >
    header1
  </h1>
  <h2
    id="header1-header2"
  >
    header2
  </h2>
  <h3
    id="header1-header2-header3"
  >
    header3
  </h3>
  <h4
    id="header1-header2-header3-header4"
  >
    header4
  </h4>
</div>
`;

exports[`Two step render result should be same with marksy html 1`] = `
<div>
  <div>
    <div>
      hello
    </div>
  </div>
</div>
`;

exports[`Two step render result should be same with marksy image 1`] = `
<div>
  <p>
    <img
      alt="test"
      src="http://some.com/image.png"
    />
  </p>
</div>
`;

exports[`Two step render result should be same with marksy italic text 1`] = `
<div>
  <p>
    hello 
    <em>
      there
    </em>
  </p>
</div>
`;

exports[`Two step render result should be same with marksy links 1`] = `
<div>
  <p>
    <a
      href="http://example.com"
    >
      my link
    </a>
  </p>
</div>
`;

exports[`Two step render result should be same with marksy list 1`] = `
<div>
  <ul>
    <li>
      foo
    </li>
    <li>
      bar
    </li>
  </ul>
</div>
`;

exports[`Two step render result should be same with marksy multiple html 1`] = `
<div>
  <div>
    <div>
      hello
    </div>
    

    <strong>
      there
    </strong>
    

    <em>
      world
    </em>
  </div>
</div>
`;

exports[`Two step render result should be same with marksy nested lists 1`] = `
<div>
  <ul>
    <li>
      Colors
      <ul>
        <li>
          Red
        </li>
        <li>
          Blue
        </li>
      </ul>
    </li>
    <li>
      Shape
      <ul>
        <li>
          Triangle
        </li>
        <li>
          Rectangle
        </li>
      </ul>
    </li>
  </ul>
</div>
`;

exports[`Two step render result should be same with marksy ordered list 1`] = `
<div>
  <ol>
    <li>
      foo
    </li>
    <li>
      bar
    </li>
  </ol>
</div>
`;

exports[`Two step render result should be same with marksy self-closing tag 1`] = `
<div>
  <p>
      
    <img
      alt="test"
      src="http://some.com/image.png"
    />
  </p>
  <div>
      
    <div>
      <br />
    </div>
    


  </div>
  <div>
      
    <hr />
    


  </div>
  <div>
      
    <input
      type="text"
    />
    

  </div>
</div>
`;

exports[`Two step render result should be same with marksy should allow injecting context to elements 1`] = `
<div>
  <div>
    bar
  </div>
</div>
`;

exports[`Two step render result should be same with marksy should allow overriding block code element 1`] = `
<div>
  <div>
    js
    :
    code
  </div>
</div>
`;

exports[`Two step render result should be same with marksy should allow overriding inline code element 1`] = `
<div>
  <p>
    Hello 
    <span>
      code
    </span>
  </p>
</div>
`;

exports[`Two step render result should be same with marksy should be able to combine in compilation 1`] = `
<div>
  <h1
    id="hey"
  >
    hey
  </h1>
  <ul>
    <li>
      foo
    </li>
  </ul>
  <p>
    -bar
  </p>
</div>
`;

exports[`Two step render result should be same with marksy should escape code when no highlighting is supplied 1`] = `
<div>
  <pre>
    <code
      class="language-js"
    >
      const Foo = () =&gt; &lt;div/&gt;
    </code>
  </pre>
</div>
`;

exports[`Two step render result should be same with marksy should handle same name nested headers 1`] = `
<div>
  <h1
    id="header1"
  >
    header1
  </h1>
  <h2
    id="header1-header2"
  >
    header2
  </h2>
  <h1
    id="header3"
  >
    header3
  </h1>
  <h2
    id="header3-header2"
  >
    header2
  </h2>
</div>
`;

exports[`Two step render result should be same with marksy should produce TOC 1`] = `
<div>
  <h1
    id="foo"
  >
    foo
  </h1>
  <h2
    id="foo-bar"
  >
    bar
  </h2>
  <h3
    id="foo-bar-baz"
  >
    baz
  </h3>
</div>
`;

exports[`Two step render result should be same with marksy should produce custom tags 1`] = `
<div>
  <div>
    foo
  </div>
</div>
`;

exports[`Two step render result should be same with marksy should work with Inferno 1`] = `"<div><div>foo</div></div>"`;

exports[`Two step render result should be same with marksy should work with Preact 1`] = `"<div><div>foo</div></div>"`;

exports[`Two step render result should be same with marksy strong text 1`] = `
<div>
  <p>
    hello 
    <strong>
      there
    </strong>
  </p>
</div>
`;

exports[`Two step render result should be same with marksy tables 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th>
          Tables
        </th>
        <th
          class="text-center"
        >
          Are
        </th>
        <th
          class="text-right"
        >
          Cool
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          col 3 is
        </td>
        <td
          class="text-center"
        >
          right-aligned
        </td>
        <td
          class="text-right"
        >
          $1600
        </td>
      </tr>
      <tr>
        <td>
          col 2 is
        </td>
        <td
          class="text-center"
        >
          centered
        </td>
        <td
          class="text-right"
        >
          $12
        </td>
      </tr>
      <tr>
        <td>
          zebra stripes
        </td>
        <td
          class="text-center"
        >
          are neat
        </td>
        <td
          class="text-right"
        >
          $1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Two step render result should be same with marksy text 1`] = `
<div>
  <p>
    hello
  </p>
</div>
`;
